test/dynamo/test_repros.py

```diff
@@ -5752,6 +5752,42 @@ def fn(x):
 
         fn(torch.randn(4))
 
+    # https://github.com/pytorch/pytorch/issues/88813
+    def test_return_value_duplication_grad(self) -> None:
+        def fn(val: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+            return val * 2, val * 2
```

Comment:
What happens if we have:
```py
def fn(val: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
    x = val * 2
    y = val * 2
    return x[0], y[0]
```
I'm worried that CSE will change the graph into:
```py
def fn(val: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
    x = val * 2
    return x[0], x[0]
```
which then has the same problem

Reply:
@zou3519 Ran that using `TORCH_COMPILE_DEBUG=1`; the FX graph looks like what we want:

```python
class GraphModule(torch.nn.Module):
    def forward(self, primals_1: "f32[2]"):
         # File: /home/bglass/git/pytorch/test/dynamo/test_repros.py:5758 in fn, code: x = val * 2
        mul: "f32[2]" = torch.ops.aten.mul.Tensor(primals_1, 2);  primals_1 = None
        
         # File: /home/bglass/git/pytorch/test/dynamo/test_repros.py:5760 in fn, code: return x[0], y[0]
        select: "f32[]" = torch.ops.aten.select.int(mul, 0, 0)
        select_1: "f32[]" = torch.ops.aten.select.int(mul, 0, 0);  mul = None
        return (select, select_1)
```

However, you do appear to be correct once we translate into IR:

```
op0: SchedulerNode(ComputedBuffer)
op0.writes = [MemoryDep('buf0', c0, {c0: 2}, None)]
op0.unmet_dependencies = []
op0.met_dependencies = [MemoryDep('primals_1', c0, {c0: 2}, None)]
op0.outputs = [
    buf0: ComputedBuffer
    buf0.layout = FixedLayout('cpu', torch.float32, size=[2], stride=[1])
    buf0.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
]
op0.group.device = cpu
op0.group.iteration = ((2,), ())
op0.sizes = ([2], [])
primals_1_layout = FixedLayout('cpu', torch.float32, size=[2], stride=[1])
buf0_layout = FixedLayout('cpu', torch.float32, size=[2], stride=[1])
class op0_loop_body:
    var_ranges = {z0: 2}
    index0 = z0
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('primals_1', get_index)
        constant = ops.constant(2.0, torch.float32)
        mul = ops.mul(load, constant)
        get_index_1 = self.get_index('index0')
        store = ops.store('buf0', get_index_1, mul, None)
        return store
```

Reply:
The FX graph isn't what we want, because now the returns share memory

Reply:
Ach, good point. Hmm. I'll keep digging here, but I'm open to suggestions on this if anyone has good ideas.